segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/exts3/EXTS3StorageTest.java

```diff
@@ -0,0 +1,281 @@
+                    }
+                }
+                int bytesRead = ( (InputStream) content).read(totalByes, Math.toIntExact(range.getFirst()),
+                        Math.toIntExact(range.getLast() + 1 - range.getFirst()));
+
+                if ( bytesRead != range.getLast() + 1 - range.getFirst()) {
+                    throw new IllegalArgumentException();
+                }
+                super.putObject( new PutObjectRequest(bucketName, key, (Object) new ByteArrayInputStream(totalByes)));
+                ACL_MAP.get(key).setSize(range.getLast() -1);
+            } catch (IOException e) {
+            }
+        }
+
+        @Override
+        public void setObjectAcl(String bucketName, String key, AccessControlList acl) {
+            AclSize retVal = ACL_MAP.get(key);
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 500, "NoSuchKey", key);
+            }
+            retVal.setAcl(acl);
+        }
+
+        @Override
+        public void setObjectAcl(SetObjectAclRequest request) {
+            AclSize retVal = ACL_MAP.get(request.getKey());
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 500, "NoSuchKey", request.getKey());
+            }
+            retVal.setAcl(request.getAcl());
+        }
+
+        @Override
+        public AccessControlList getObjectAcl(String bucketName, String key) {
+            AclSize retVal = ACL_MAP.get(key);
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 500, "NoSuchKey", key);
+            }
+            return retVal.getAcl();
+        }
+
+        public CopyPartResult copyPart(CopyPartRequest request) {
+            if ( ACL_MAP.get(request.getKey()) == null ) {
+                throw new S3Exception("NoObject", 500, "NoSuchKey", request.getKey());
+            }
+
+            Range range = request.getSourceRange();
+            if ( range.getLast() == -1 ) {
+                range = Range.fromOffsetLength(0, ACL_MAP.get(request.getSourceKey()).getSize());
+                request.withSourceRange(range);
+            }
+            CopyObjectResult result = executeRequest(client, request, CopyObjectResult.class);
+            CopyPartResult retVal = new CopyPartResult();
+            retVal.setPartNumber(request.getPartNumber());
+            retVal.setETag(result.getETag());
+            return retVal;
+        }
+
+        @Override
+        public void deleteObject(String bucketName, String key) {
+            super.deleteObject(bucketName, key);
+            ACL_MAP.remove(key);
+        }
+
+        @Override
+        public DeleteObjectsResult deleteObjects(DeleteObjectsRequest request) {
+            request.getDeleteObjects().getKeys().forEach( (key) -> ACL_MAP.remove(key));
+            return super.deleteObjects(request);
+        }
+    }
+
+    @Data
```

Comment:
Using @Wither and @Synchronized. Though I think just using @Synchronized would have solved.